ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java

```diff
@@ -196,6 +198,11 @@ private void handleStatementWithTerminatedQueries(
       executeStatement(
           statement, command, commandId, commandStatusFuture, mode);
     } catch (final KsqlException exception) {
+      final Throwable rootCause = ExceptionUtils.getRootCause(exception);
+      if (mode == Mode.RESTORE && rootCause instanceof KsqlMissingSourceException) {
```

Comment:
Apologies if I've missed something as I'm late to the game here: there are multiple reasons a statement could during recovery, besides a missing source, so I don't believe this approach works. (For example, I originally discovered the recovery bug when a statement failed to execute because Schema Registry rejected a sink schema due to a failed schema compatibility check.)

If we're pursuing this fix for legacy query IDs (rather than updating how query IDs are generated and leaving the bug for legacy query IDs, as proposed in the discussion above), then we should always increment the query ID on recovery, regardless of the reason for failure.